core/state/database.go

```diff
@@ -543,32 +546,55 @@ func (db *cachingDB) SetLastMerkleRoot(merkleRoot common.Hash) {
 }
 
 func (db *cachingDB) SaveTransitionState(root common.Hash) {
-	if db.TransitionStatePerRoot == nil {
-		db.TransitionStatePerRoot = make(map[common.Hash]*TransitionState)
-	}
+	db.TransitionStatePerRoot = lru.NewBasicLRU[common.Hash, *TransitionState](100)
 
 	if db.CurrentTransitionState != nil {
+		encoded, err := rlp.EncodeToBytes(db.CurrentTransitionState)
```

Comment:
I should not try to write to the db unless the value wasn't present in the LRU. This is going to add too many writes otherwise, since upon creation the conversion can be processed many times.